packages/menu/src/MenuItem.ts

```diff
@@ -146,41 +189,37 @@ export class MenuItem extends ActionButton {
         }
     }
 
+    public connectedCallback(): void {
+        super.connectedCallback();
+        addOrUpdateEvent.reset(this);
+        this.dispatchEvent(addOrUpdateEvent);
+    }
+
     public disconnectedCallback(): void {
         super.disconnectedCallback();
-        const removedEvent = new CustomEvent('sp-menu-item-removed', {
-            bubbles: true,
-            composed: true,
-            detail: {
-                item: this,
-            },
-        });
-        this.dispatchEvent(removedEvent);
+        removeEvent.reset();
+        this.dispatchEvent(removeEvent);
     }
 
     public async triggerUpdate(): Promise<void> {
         await new Promise((ready) => requestAnimationFrame(ready));
-        const updatedEvent = new CustomEvent('sp-menu-item-update', {
-            bubbles: true,
-            composed: true,
-            detail: {
-                item: this,
-                owned: false,
-            },
-        });
-        this.dispatchEvent(updatedEvent);
+        addOrUpdateEvent.reset(this);
+        this.dispatchEvent(addOrUpdateEvent);
     }
-}
 
-export interface MenuItemUpdateEvent {
-    item: MenuItem;
-    owned: boolean;
+    public menuData: {
+        focusRoot?: Menu;
+        selectionRoot?: Menu;
+    } = {
+        focusRoot: undefined,
+        selectionRoot: undefined,
+    };
 }
 
 declare global {
     interface GlobalEventHandlersEventMap {
-        'sp-menu-item-added': CustomEvent<MenuItemUpdateEvent>;
-        'sp-menu-item-update': CustomEvent<MenuItemUpdateEvent>;
-        'sp-menu-item-removed': CustomEvent<MenuItemUpdateEvent>;
+        'sp-menu-item-added': MenuItemAddedOrUpdatedEvent;
+        'sp-menu-item-update': MenuItemAddedOrUpdatedEvent;
```

Comment:
This isn't actually dispatched anymore. I also think you should change the added event to `sp-menu-item-added-or-updated` (or re-work things a bit so both can get dispatched separately).